### PR TITLE
Issue 180 Adding contact-direct check-in

### LINKED ISSUE #102: Email reminders
### User Story
As an entity
I'd like to receive email alerts at defined intervals 
So that I can be reminded to check-in with the city.

### Acceptance Criteria
- [x] Functionality to send emails as an alert
- [ ] PR merged to `master`

### LINKED ISSUE #131: Admin Roles Access
### User Story
As an Admin
I need the API to apply access
So that I can that I can perform a function

### Acceptance Criteria
- [x] Verify that the Entity is able to POST, GET, PUT and DELETE
- [x] Verify that the User is able to POST, GET, PUT and DELETE
- [x] Verify that the Contact is able to POST, GET, PUT and DELETE
- [x] PR merged to `master`

### Notes
GET - get all roles
POST - add a role
DELETE - Add a role  (POST via "/delete")

### LINKED ISSUE #132: City/Outreach User Roles
### User Story
As a City/Outreach User,
I need the API to apply access
So that I can perform a function

### Acceptance Criteria
- [x] Verify that the Entity is able to POST, GET and PUT
- [x] Verify that the User is able to GET and PUT
- [x] Verify that the Contact is able to POST, GET and PUT
- [x] Verify that the Entity, User, and Contact are not permitted to DELETE
- [x] Verify that the User is not permitted to POST
- [x] Verify that the User is not permitted to get a CSV file of Entities, Contact, or Users

### Notes
POST - Add a role
GET - Get all roles
DELETE - delete a role (POST via "/delete")

### LINKED ISSUE #145: System unable to load the New Check-In questions when user create an Entity through the API.
### User Story
As a user of Bmore responsive
I would like for the type column to be populated correctly
so that the check in questions associated with that type are available in the HealthCare UI

### Acceptance Criteria
- [x] Type is populated when an entity is created or updated
- [x] Swagger doc is updated
- [ ] PR merged to `master`

### Notes
While user create an Entity using API, look like type column is not updated in the entity table. 

image

### LINKED ISSUE #158: UserRole Export to CSV is not returning results
### User Story
As a developer on the Bmore-responsive project
I'd like to be able to export UserRoles to csv
So that I can see which UserRoles are available to create/update users with

### Acceptance Criteria
- [x] Remove UserRoles and User exports
- [x] Swagger doc is updated
- [x] Postman collection is updated if required
- [ ] PR merged to `master`

### Notes
When you try to export UserRole to CSV, no results are returned.  Here is the messaging from the local window running bmore-responsive:

Executing (default): SELECT "id", "ptype", "v0", "v1", "v2", "v3", "v4", "v5", "createdAt", "updatedAt" FROM "public"."casbin_rules" AS "casbin_rule";
::1 - - [20/May/2020:11:47:56 +0000] "GET /csv/UserRole HTTP/1.1" 200 -

### LINKED ISSUE #162: update entity failed when populated based on swagger doc - issue with contacts and checkins with update
### User Story
As a user of the bmore-responsive system
I'd like to be able to update an entity based on the swagger doc
So that I can update entities quickly and efficiently

### Acceptance Criteria
- [x] Functionality of system and expected functionality of system is reflected in swagger doc
- ~~[ ] an entity can be updated with check-ins~~
- ~~[ ] an entity can be updated with contacts~~
- [x] swagger doc is updated
- [x] postman collection is updated
- [ ] PR merged to `master`

### Notes
I tried to update an entity based on what is in swagger doc.  I had to remove the check-in data and contact data in order to update an entity.

### LINKED ISSUE #173: can create a user with an empty password - password required to log in
**Describe the bug**
A user can be created with an empty password but you cannot log into HealthCare RollCall UI with an empty password

**To Reproduce**
1. Create a user using Post url/user with empty password ""

**Expected behavior**
A password should probably be required and the user should not be created without a password.

### LINKED ISSUE #180: Provider-direct check-in
### User Story
As a entity contact
I'd like to be able to check-in directly 
So that I can update my status without speaking with anyone directly

### Acceptance Criteria
- [x] temp user tokens generated for contacts
- [x] contact can access their associated entity via token
- [x] endpoint to trigger email to all contacts with token and link to check-in.

### Notes
This will require work on the front-end as well.

### PR DESCRIPTION
**Pull request by revjtanton**
This provides a variety of bug fixes as defined by the resolved tickets.
Primarily this PR will add contact-direct check-in capabilities. This allows
Authorized contacts to check-in entities directly without a user account by using an
auth token.
resolves #131
resolves #132
resolves #102
resolves #162
resolves #173
resolves #145
resolves #158
resolves #180 

## Todos

- [x] Tests
- [x] Documentation

## Deploy Notes

There is a new endpoint defined in Swagger that will allow for emails to be sent to contacts. Sending that email blast is touchy. To test you must set an `.env` var for: -
- `URL=http://localhost:8080` an
- `TEST_EMAIL=yourmail@codeforbaltimore.org`

Then you must make sure at least one contact email in the system is equal to your test email set above.

### CODE REVIEW

File: src/models/entity-contact.js [hunk 1]
```diff
@@ -48,6 +48,14 @@ const entityContact = (sequelize, DataTypes) => {
         }
     }
 
+    EntityContact.findByEntityId = async (entityId) => {
+        const entries = await EntityContact.findAll({
+            where: {entityId}
```
Comment (c-w-allen) [inline src/models/entity-contact.js:53]: Does this automatically find it by pk?

File: src/routes/contact.js [hunk 2]
```diff
@@ -92,6 +93,54 @@ router.post('/', async (req, res) => {
...
+router.post('/send', async (req, res) => {
+	let code;
+	let message;
+	const emails = [];
+
+	try {
+		/** @todo allow for passing entity and contact arrays */
+		const { entityIds, contactIds, relationshipTitle } = req.body;
+		
+		if (entityIds === undefined && contactIds === undefined) {
```
Comment (c-w-allen) [inline src/routes/contact.js:106]: It looks like this is only running if `entityIds` and `contactIds` is undefined. If that's the case. are they needed in the request body?

File: src/routes/user.js [hunk 3]
```diff
@@ -146,7 +147,8 @@ router.post('/', utils.authMiddleware, async (req, res) => {
 	let code;
 	let message;
 	try {
-		if (validator.isEmail(req.body.email)) {
+		console.log(req.body.password)
```
Comment (c-w-allen) [inline src/routes/user.js:150]: Is this safe to keep?

File: src/routes/contact.js (hunk 2 above)
Comment (revjtanton) [reply src/routes/contact.js:106]: That's for future planning. Eventually, we're going to allow for passing an array of either or both types of ID's to _only_ send emails to those associations. For now, though, that functionality isn't in there.

File: src/models/entity-contact.js (hunk 1 above)
Comment (revjtanton) [reply src/models/entity-contact.js:53]: It does not for some reason. Not sure what was off about that but it wasn't really worth debugging. When I used `findByPk` nothing came back.

File: src/routes/user.js (hunk 3 above)
Comment (revjtanton) [reply src/routes/user.js:150]: no that probably needs to go 😎

**Comment by revjtanton** [comment]
@allcontributors add @c-w-allen for review

**Comment by revjtanton** [comment]
@allcontributors add @msmith125 for review

**Pull request merged**